### PR TITLE
Add sticker manager bot

### PR DESCRIPTION
**Pull request by tulir**
Missing things:
* [x] Limiting to one poll per day
* [x] Close polls after a day if they don't get enough votes
* [x] Close polls if they get too many downvotes

Future enhancements
* Allow adding animated stickers to separate pack

### CODE REVIEW

File: stickermanager.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+sticker_pack = None
+
+with open(os.path.join(os.path.dirname(__file__), 'stickermanager.tsv')) as f:
+    WEIGHTS = {
+        int(uid) if uid.strip().isnumeric() else uid:
+            (0 if weight == '-' else float(weight), None if name == '-' else name)
+        for uid, weight, name
+        in (line.strip().split('\t', 2)
+            for line in f
+            if len(line) > 0 and not line[0] == '#' and not line.isspace())
```
Comment (Lonami) [inline stickermanager.py]: Python prefers using truthy values (`if line`) when possible.

Comment (Lonami) [reply stickermanager.py]: Normally `line.startswith` makes the intention clearer, but it's okay to leave the check against the first character.

File: stickermanager.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    }
+
+    VOTES_REQUIRED = WEIGHTS.pop('votes required')[0]
+    DEFAULT_WEIGHT = WEIGHTS.pop('default weight')[0]
+    STICKER_PACK_TITLE = WEIGHTS.pop('sticker title')[1]
+    STICKER_PACK_SHORT_NAME = WEIGHTS.pop('sticker short')[1]
+    ADMIN_USER_ID = int(WEIGHTS.pop('sticker owner')[1])
+    ALLOWED_CHATS = [int(gid.strip())
```
Comment (Lonami) [inline stickermanager.py]: The `strip()` call is not necessary. `int()` already ignores leading and trailing whitespace.

File: stickermanager.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+                     for gid in WEIGHTS.pop('allowed chats')[1].split(',')
+                     if not gid.isspace()]
+
+
+async def create_sticker_pack(bot: TelegramClient, item: InputStickerSetItem
+                              ) -> Tuple[bool, Optional[StickerSet]]:
+    global sticker_pack
+    try:
+        with open(os.path.join(os.path.dirname(__file__), 'stickermanager.json')) as file:
```
Comment (Lonami) [inline stickermanager.py]: re: `pathlib.Path` consideration.

File: stickermanager.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+    except (FileNotFoundError, KeyError) as e:
+        stickerset: StickerSet = await bot(CreateStickerSetRequest(
+            user_id=ADMIN_USER_ID,
+            title=STICKER_PACK_TITLE,
+            short_name=STICKER_PACK_SHORT_NAME,
+            stickers=[item]
+        ))
+        sticker_pack = InputStickerSetID(id=stickerset.set.id,
+                                         access_hash=stickerset.set.access_hash)
+        with open(os.path.join(os.path.dirname(__file__), 'stickermanager.json'), 'w') as file:
```
Comment (Lonami) [inline stickermanager.py]: re: `pathlib.Path` consideration.

File: stickermanager.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+    if animated:
+        file = await bot.upload_file(current_vote['filepath'])
+        mime = 'application/x-tgsticker'
+    else:
+        img = Image.open(current_vote['filepath'])
+        w, h = img.size
+        if w > 512 or h > 512:
+            img.thumbnail((512, 512), Image.ANTIALIAS)
+        else:
+            if w > h:
```
Comment (Lonami) [inline stickermanager.py]: I personally prefer to merge this into an `elif`.

File: stickermanager.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+                img = img.resize((512, int(h * (512 / w))), Image.ANTIALIAS)
+            elif w < h:
+                img = img.resize((int((w * (512 / h))), 512), Image.ANTIALIAS)
+            else:
+                img = img.resize((512, 512), Image.ANTIALIAS)
```
Comment (Lonami) [inline stickermanager.py]: This is the same case as the first `if`. Probably the `w > h` and `w < h` branches can go first, *and then* have this be the last one.

File: stickermanager.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+            _, sender_name = WEIGHTS[event.sender_id]
+        except KeyError:
+            return
+        orig_evt: Message = await event.get_reply_message()
+        # TODO add support for animated stickers
+        if not orig_evt.photo and (not orig_evt.sticker or
+                                   orig_evt.sticker.mime_type == 'application/x-tgsticker'):
+            return
+
+        filename = os.path.join(os.path.dirname(__file__), f'stickermanager.{int(time())}.dat')
```
Comment (Lonami) [inline stickermanager.py]: re: `pathlib.Path` suggestion.

File: stickermanager.py [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+            'score': 0,
+            'emoji': emoji,
+            'votes': {},
+            'filepath': filename,
+            'pack_link': f'https://t.me/addstickers/{STICKER_PACK_SHORT_NAME}',
+            'animated': orig_evt.sticker and orig_evt.sticker.mime_type == 'application/x-tgsticker'
+        }
+        reply_evt: Message = await orig_evt.reply(
+            POLL_TEMPLATE.format_map(get_template_data()),
+            buttons=[Button.inline(UP, b'+'), Button.inline(DOWN, b'-')],
```
Comment (Lonami) [inline stickermanager.py]: This button data may be easily confused in other buttons. Perhaps prefixing it (e.g. `b'addsticker/+'`) is a good idea?

File: stickermanager.py [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+        orig_evt: Message = await event.get_reply_message()
+        # TODO add support for animated stickers
+        if not orig_evt.photo and (not orig_evt.sticker or
+                                   orig_evt.sticker.mime_type == 'application/x-tgsticker'):
+            return
+
+        filename = os.path.join(os.path.dirname(__file__), f'stickermanager.{int(time())}.dat')
+        await orig_evt.download_media(filename)
+
+        asyncio.ensure_future(event.delete(), loop=bot.loop)
```
Comment (Lonami) [inline stickermanager.py]: Please note that creating a future or task and never `await`'ing it is discouraged. Consider moving this inside an `asyncio.wait/gather` or simply `await` it here.

File: stickermanager.py [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+            'filepath': filename,
+            'pack_link': f'https://t.me/addstickers/{STICKER_PACK_SHORT_NAME}',
+            'animated': orig_evt.sticker and orig_evt.sticker.mime_type == 'application/x-tgsticker'
+        }
+        reply_evt: Message = await orig_evt.reply(
+            POLL_TEMPLATE.format_map(get_template_data()),
+            buttons=[Button.inline(UP, b'+'), Button.inline(DOWN, b'-')],
+            parse_mode='html')
+        current_vote['poll'] = reply_evt.id
+        asyncio.ensure_future(reply_evt.pin(notify=True), loop=bot.loop)
```
Comment (Lonami) [inline stickermanager.py]: re: not `await`'ing the `Future`.

Comment (Lonami) [reply stickermanager.py]: Is the "pinotify" necessary?

File: stickermanager.py [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+
+    @bot.on(events.CallbackQuery(chats=ALLOWED_CHATS, data=lambda data: data in (b'+', b'-')))
```
Comment (Lonami) [inline stickermanager.py]: You should be able to use a regex pattern for the data.

File: stickermanager.py [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+    async def vote_poll(event: events.CallbackQuery.Event) -> None:
+        if not current_vote or current_vote['poll'] != event.message_id:
+            await event.answer('That poll is closed.')
+            return
+
+        async with current_vote_lock:
```
Comment (Lonami) [inline stickermanager.py]: The use of this lock seems error prone. Perhaps it should be abstracted into its own function (e.g. `locked_vote_poll_int`)? Or better, make `vote_poll_int` private or even local. Or better, use a `@locked(current_vote_lock)` decorator.

File: stickermanager.py [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+            return
+
+        weight, displayname = WEIGHTS.get(event.sender_id, (DEFAULT_WEIGHT, None))
+        if weight == 0:
+            await event.answer('You don\'t have the permission to vote.')
+            return
+        if event.data == b'-':
+            weight = -weight
+
+        displayname = displayname or escape_html((await event.get_sender()).first_name)
```
Comment (Lonami) [inline stickermanager.py]: Note that `utils.get_display_name` exists, but perhaps using only the first name is more appropriated.

File: stickermanager.py [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+                res = "accepted"
+            else:
+                current_vote['result'] = RESULT_REJECTED.format_map(get_template_data())
+                res = "rejected"
+            await bot.edit_message(current_vote['chat'], current_vote['poll'],
+                                   POLL_FINISHED_TEMPLATE.format_map(get_template_data()),
+                                   parse_mode='html')
+            await event.answer(f'Successfully voted {fancy_round(weight)},'
+                               f' which made the sticker be {res} \U0001f389')
+            if res == "accepted":
```
Comment (Lonami) [inline stickermanager.py]: Given `res` only seems to be used here, and checked against a value with two possible outcomes, wouldn't it be better to have `accepted = scores.sum > 0`?

File: stickermanager.py [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+                pack, document = await add_sticker_to_pack(bot)
+                await event.respond(file=document, reply_to=current_vote['poll'])
+            current_vote = None
+        else:
+            await bot.edit_message(current_vote['chat'], current_vote['poll'],
+                                   POLL_TEMPLATE.format_map(get_template_data()),
+                                   buttons=[Button.inline(f'{UP} ({scores.yes})', b'+'),
```
Comment (Lonami) [inline stickermanager.py]: re: buttons' data.

File: stickermanager.tsv [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+# The vote weight for users who aren't listed.
+default weight	0.1	-
+
+# Comma-separated list of chat IDs where the bot is enabled.
+allowed chats	-	-1001200633650
+
+# Users
+# Format:
+# user id	weight	name
+10885151 	10	lonami
```
Comment (Lonami) [inline stickermanager.tsv]: This is truly unfair.

File: stickermanager.py (hunk 10 above)
Comment (tulir) [reply stickermanager.py]: notify is definitely necessary :D

File: stickermanager.py (hunk 11 above)
Comment (tulir) [reply stickermanager.py]: is regex really better than just using `in` when there are exactly two options?

File: stickermanager.py (hunk 13 above)
Comment (tulir) [reply stickermanager.py]: Yeah, people in the whitelist also only have first names

File: stickermanager.py [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+DOWN = '\U0001f53d'
+
+VoteData = NamedTuple('VoteData', weight=int, displayname=str)
+Scores = NamedTuple('Scores', sum=int, yes=int, no=int)
+
+current_vote: Optional[dict] = None
+current_vote_lock: asyncio.Lock = asyncio.Lock()
+sticker_pack = None
+
+with open(os.path.join(os.path.dirname(__file__), 'stickermanager.tsv')) as f:
```
[the remark this answers is not in the document]
Comment (tulir) [reply stickermanager.py]: > Any reason to not use the (objectively superior) `pathlib.Path`?

autoadmin doesn't use it

File: stickermanager.tsv (hunk 16 above)
Comment (tulir) [reply stickermanager.tsv]: Now it's even more unfair

File: stickermanager.py (hunk 13 above)
Comment (YouTwitFace) [reply stickermanager.py]: I don't 😏

Comment (tulir) [reply stickermanager.py]: I mean the whitelist only has first names, you're "twit"

File: stickermanager.py [hunk 18]
```diff
@@ -0,0 +1,354 @@
...
+    global sticker_pack, current_vote, last_accepted
+    with open(CACHE_FILE) as file:
+        data = json.load(file)
+        sp_data = data['sticker_pack']
+        if sp_data:
+            sticker_pack = InputStickerSetID(id=sp_data['id'], access_hash=sp_data['access_hash'])
+        cv_data = data['current_vote']
+        if cv_data:
+            current_vote = cv_data
+            current_vote["votes"] = {int(uid): VoteData(*data)
```
Comment (Lonami) [inline stickermanager.py]: How do you judge whether to use double or single quotes? Given you made use of single quotes everywhere else (even when a string already used them, requiring escaping), shouldn't this be:
```suggestion
            current_vote['votes'] = {int(uid): VoteData(*data)
```

File: stickermanager.py [hunk 19]
```diff
@@ -0,0 +1,354 @@
...
+                                     for uid, data in cv_data["votes"].items()}
```
Comment (Lonami) [inline stickermanager.py]: re: single quotes.
```suggestion
                                     for uid, data in cv_data['votes'].items()}
```

File: stickermanager.py [hunk 20]
```diff
@@ -0,0 +1,354 @@
...
+
+    async def _locked_finish_poll() -> bool:
+        global current_vote, last_accepted
+
+        if not current_vote:
+            return False
+
+        unpin_task = asyncio.ensure_future(bot.pin_message(current_vote['chat'], message=None),
+                                           loop=bot.loop)
+        accepted = current_vote["score"] >= VOTES_REQUIRED
```
Comment (Lonami) [inline stickermanager.py]: re: single quotes
```suggestion
        accepted = current_vote['score'] >= VOTES_REQUIRED
```

File: stickermanager.py [hunk 21]
```diff
@@ -0,0 +1,354 @@
...
+
+        current_vote['votes'][event.sender_id] = VoteData(weight=weight, displayname=displayname)
+
+        scores = calculate_scores()
+        current_vote['score'] = scores.sum
+
+        if abs(scores.sum) >= VOTES_REQUIRED:
+            current_vote_status.set()
+            accepted = await _locked_finish_poll()
+            res = "accepted" if accepted else "rejected"
```
Comment (Lonami) [inline stickermanager.py]: re: single quotes
```suggestion
            res = 'accepted' if accepted else 'rejected'
```

File: stickermanager.py [hunk 22]
```diff
@@ -0,0 +1,354 @@
...
+
+UP = '\U0001f53c'
+DOWN = '\U0001f53d'
+UP_DAT = b'addsticker/+'
+DOWN_DAT = b'addsticker/-'
+
+POLL_TIMEOUT = ADD_COOLDOWN = 24 * 60 * 60
+
+BASE_DIR = Path(__file__).resolve().parent
+CONFIG_FILE = BASE_DIR.joinpath("stickermanager.tsv")
```
Comment (Lonami) [inline stickermanager.py]:
```suggestion
CONFIG_FILE = BASE_DIR / 'stickermanager.tsv'
```

File: stickermanager.py [hunk 23]
```diff
@@ -0,0 +1,354 @@
...
+CACHE_FILE = BASE_DIR.joinpath("stickermanager.json")
```
Comment (Lonami) [inline stickermanager.py]:
```suggestion
CACHE_FILE = BASE_DIR / 'stickermanager.json'
```

File: stickermanager.py [hunk 24]
```diff
@@ -0,0 +1,354 @@
...
+DATA_FILE_FORMAT: str = "stickermanager.{ts}.dat"
```
Comment (Lonami) [inline stickermanager.py]:
```suggestion
DATA_FILE_FORMAT: str = 'stickermanager.{ts}.dat'
```

File: stickermanager.py [hunk 25]
```diff
@@ -0,0 +1,354 @@
...
+            return
+        elif current_vote:
+            await event.reply('There\'s already an ongoing sticker poll.')
+            return
+
+        async with current_vote_lock:
+            await _locked_start_poll(event)
+            save_cache()
+
+    async def _locked_start_poll(event: Union[events.NewMessage.Event, Message]) -> None:
```
Comment (Lonami) [inline stickermanager.py]: …this is not what I intended. The following is:

```python
async def locked_start_poll(event: Union[events.NewMessage.Event, Message]) -> None:
    async with current_vote_lock:
        await _start_poll()
        save_cache()

async def _start_poll(event: Union[events.NewMessage.Event, Message]) -> None:
    ...
```

As it stands, the renaming is pointless.

File: stickermanager.py (hunk 18 above)
Comment (tulir) [reply stickermanager.py]: I used double quotes everywhere and then did `s/"/'/g/` because you use single quotes

**Pull request merged**